src/python/grpcio/grpc/_cython/_cygrpc/aio/channel.pyx.pxi

```diff
@@ -12,19 +12,80 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
+class _WatchConnectivityFailed(Exception):
+    """Dedicated exception class for watch connectivity failed.
+
+    It might be failed due to deadline exceeded, or the channel is closing.
+    """
+cdef CallbackFailureHandler _WATCH_CONNECTIVITY_FAILURE_HANDLER = CallbackFailureHandler(
+    'watch_connectivity_state',
+    'Timed out or channel closed.',
+    _WatchConnectivityFailed)
+
+
 cdef class AioChannel:
     def __cinit__(self, bytes target):
         self.channel = grpc_insecure_channel_create(<char *>target, NULL, NULL)
         self.cq = CallbackCompletionQueue()
         self._target = target
+        self._loop = asyncio.get_event_loop()
+        self._status = AIO_CHANNEL_STATUS_READY
 
     def __repr__(self):
         class_name = self.__class__.__name__
         id_ = id(self)
         return f"<{class_name} {id_}>"
 
+    def check_connectivity_state(self, bint try_to_connect):
+        """A Cython wrapper for Core's check connectivity state API."""
+        return grpc_channel_check_connectivity_state(
+            self.channel,
+            try_to_connect,
+        )
+
+    async def watch_connectivity_state(self,
+                                       grpc_connectivity_state last_observed_state,
+                                       object deadline):
+        """Watch for one connectivity state change.
+
+        Keeps mirroring the behavior from Core, so we can easily switch to
+        other design of API if necessary.
+        """
+        if self._status == AIO_CHANNEL_STATUS_DESTROYED:
+            # TODO(lidiz) switch to UsageError
+            raise RuntimeError('Channel is closed.')
+        cdef gpr_timespec c_deadline = _timespec_from_time(deadline)
+
+        cdef object future = self._loop.create_future()
+        cdef CallbackWrapper wrapper = CallbackWrapper(
+            future,
+            _WATCH_CONNECTIVITY_FAILURE_HANDLER)
+        cpython.Py_INCREF(wrapper)
+        grpc_channel_watch_connectivity_state(
+            self.channel,
+            last_observed_state,
+            c_deadline,
+            self.cq.c_ptr(),
+            wrapper.c_functor())
+
+        # NOTE(lidiz) The callback will be invoked after the channel is closed
+        # with a failure state. We need to keep wrapper alive until then, or we
+        # will observe a segfault.
```

Comment:
So you mean that in case of closing the `AioChannel` which would result in destructing the object, if later arrives the callback from the core for notifying a change in the satus - or a timeout - the cb would find a deallocated wrapper?

How is this aligned with this [1]? Are both trying to solve the same issue? Just saying because the description of the problem is different but they are doing basically the same, with just one little difference, for [1] if there is an exception in the `future` the `cpython.Py_DECREF(wrapper)` will never be executed which could lead to some leaks.

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi#L115

Reply:
Yes, they are caused by the same issue that the C callback function being prematurely deallocated. I updated the one in `callback_common.pyx.pxi` to the same pattern here. It is indeed a potential risk I overlooked. Good catch!

Reply:
Having the feeling that if this is the case, the `cpython.Py_DECREF(wrapper)` should be done in during the callback execution otherwise the `dealloc_wrapper` could be called just the next loop iteration after resolving the future.

With incrementing the wrapper refcount before and decrementing during the CB execution you will make sure that the object is still there when is needed.

Reply:
Great idea! I integrated the ref-count into the `CallbackWrapper` and removed all `NOTE`s around it. It can be interpreted as Core holding a reference to the wrapper object, and release the reference after the invocation of the callback.